<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Cecilia Noecker" />


<title>Multi-omics Analysis</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/spacelab.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { font-weight: bold; } /* Alert */
code span.an { font-style: italic; } /* Annotation */
code span.cf { font-weight: bold; } /* ControlFlow */
code span.co { font-style: italic; } /* Comment */
code span.cv { font-style: italic; } /* CommentVar */
code span.do { font-style: italic; } /* Documentation */
code span.dt { text-decoration: underline; } /* DataType */
code span.er { font-weight: bold; } /* Error */
code span.in { font-style: italic; } /* Information */
code span.kw { font-weight: bold; } /* Keyword */
code span.pp { font-weight: bold; } /* Preprocessor */
code span.wa { font-style: italic; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 52px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h2 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h3 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h4 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h5 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h6 {
  padding-top: 57px;
  margin-top: -57px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Multi-omics analysis tutorial, 2021</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_prep.html">Advance setup</a>
</li>
<li>
  <a href="index.html">Tutorial</a>
</li>
<li>
  <a href="full_output.html">Tutorial with full code and output</a>
</li>
<li>
  <a href="slides.pdf">Tutorial slides</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Multi-omics Analysis</h1>
<h4 class="author">Cecilia Noecker</h4>
<h4 class="date">Spring 2021</h4>

</div>


<p>This is the hands-on segment of the multi-omics analysis tutorial. The data analyzed below are described in <a href="https://www.nature.com/articles/s41564-018-0306-4">Franzosa et al Nature Microbiology, 2019</a>. Please visit the <a href="workshop_prep.html">Setup</a> page for some initial steps to complete prior to the tutorial.</p>
<div id="descriptive-analysis" class="section level1">
<h1><span class="header-section-number">1</span> Descriptive analysis</h1>
<p>What is the overall landscape of our dataset?</p>
<p>Typically, you would start the analysis of a new dataset (after data cleaning and QC) with some initial descriptive analyses to summarize the overall trends in the data across all features (taxa/genes/metabolites). The tools most commonly used for this task are similar to those you’ve already seen in Jordan’s previous session on 16S rRNA sequencing: dimensional reduction and ordination (PCA/PCoA/NMDS), visualization of feature abundances across samples (heatmaps, line plots, etc.), and statistical modeling to assess which features are differentially abundant between experimental groups. In the interest of time, we won’t work through these analyses together, but you can ordination plots based on microbiome taxa (PCoA, Bray-Curtis dissimilarity) and metabolite features (PCA) below. Some additional questions and suggestions for performing these analyses are included at the end of the document if you want to continue to explore this dataset.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">library</span>(tidymodels)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">library</span>(vegan)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">library</span>(ape)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co">## Import metadata</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7">metadata_cinc &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/subject_metadata_Cincinnati.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(data_type <span class="op">==</span><span class="st"> &quot;metabolomics&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">diagnosis =</span> <span class="kw">factor</span>(diagnosis, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;nonIBD&quot;</span>, <span class="st">&quot;UC&quot;</span>, <span class="st">&quot;CD&quot;</span>)))</a>
<a class="sourceLine" id="cb1-8" data-line-number="8">metadata_mgh &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/subject_metadata_MGH.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(data_type <span class="op">==</span><span class="st"> &quot;metabolomics&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">diagnosis =</span> <span class="kw">factor</span>(diagnosis, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;nonIBD&quot;</span>, <span class="st">&quot;UC&quot;</span>, <span class="st">&quot;CD&quot;</span>)))</a>
<a class="sourceLine" id="cb1-9" data-line-number="9"></a>
<a class="sourceLine" id="cb1-10" data-line-number="10"><span class="co">## Import metabolite data</span></a>
<a class="sourceLine" id="cb1-11" data-line-number="11">metabolites_cinc &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/metabolite_profile_Cincinnati.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-12" data-line-number="12"><span class="st">  </span><span class="kw">rename_with</span>( <span class="op">~</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;_</span><span class="ch">\\</span><span class="st">(.*&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">gsub</span>(<span class="st">&quot; |</span><span class="ch">\\</span><span class="st">/&quot;</span>, <span class="st">&quot;_&quot;</span>, .x))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-13" data-line-number="13"><span class="st">  </span><span class="kw">pivot_longer</span>(<span class="dt">names_to =</span> <span class="st">&quot;SampleID&quot;</span>, <span class="dt">cols =</span> <span class="kw">contains</span>(<span class="st">&quot;HSM&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-14" data-line-number="14"><span class="st">  </span><span class="kw">left_join</span>(<span class="kw">select</span>(metadata_cinc, SampleID, SubjectID, diagnosis), <span class="dt">by =</span> <span class="st">&quot;SampleID&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-15" data-line-number="15"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">value_zeros =</span> <span class="kw">ifelse</span>(<span class="kw">is.na</span>(value), <span class="dv">0</span>, value))</a>
<a class="sourceLine" id="cb1-16" data-line-number="16">metabolites_mgh &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/metabolite_profile_MGH.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-17" data-line-number="17"><span class="st">  </span><span class="kw">rename_with</span>( <span class="op">~</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;_</span><span class="ch">\\</span><span class="st">(.*&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">gsub</span>(<span class="st">&quot; |</span><span class="ch">\\</span><span class="st">/&quot;</span>, <span class="st">&quot;_&quot;</span>, .x))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-18" data-line-number="18"><span class="st">  </span><span class="kw">pivot_longer</span>(<span class="dt">names_to =</span> <span class="st">&quot;SampleID&quot;</span>, <span class="dt">cols =</span> <span class="kw">contains</span>(<span class="st">&quot;MSM&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-19" data-line-number="19"><span class="st">  </span><span class="kw">left_join</span>(<span class="kw">select</span>(metadata_mgh, SampleID, SubjectID, diagnosis), <span class="dt">by =</span> <span class="st">&quot;SampleID&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-20" data-line-number="20"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">value_zeros =</span> <span class="kw">ifelse</span>(<span class="kw">is.na</span>(value), <span class="dv">0</span>, value))</a>
<a class="sourceLine" id="cb1-21" data-line-number="21"></a>
<a class="sourceLine" id="cb1-22" data-line-number="22">metabolites_all &lt;-<span class="st"> </span><span class="kw">rbind</span>(<span class="kw">mutate</span>(metabolites_cinc, <span class="dt">study_site =</span> <span class="st">&quot;Cincinnati&quot;</span>), <span class="kw">mutate</span>(metabolites_mgh, <span class="dt">study_site =</span> <span class="st">&quot;MGH&quot;</span>))</a>
<a class="sourceLine" id="cb1-23" data-line-number="23"><span class="co">## Filter features present in fewer than 5 samples and log-transform values</span></a>
<a class="sourceLine" id="cb1-24" data-line-number="24">bad_feats &lt;-<span class="st"> </span>metabolites_all <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(Compound) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-25" data-line-number="25"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">numVar =</span> <span class="kw">length</span>(<span class="kw">unique</span>(value_zeros))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb1-26" data-line-number="26"><span class="st">  </span><span class="kw">filter</span>(numVar <span class="op">&lt;</span><span class="st"> </span><span class="dv">5</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(Compound)</a>
<a class="sourceLine" id="cb1-27" data-line-number="27">metabolites_all &lt;-<span class="st"> </span><span class="kw">filter</span>(metabolites_all, <span class="op">!</span>Compound <span class="op">%in%</span><span class="st"> </span>bad_feats) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">logValueZeros =</span> <span class="kw">log10</span>(value_zeros<span class="op">+</span><span class="dv">1</span>))</a>
<a class="sourceLine" id="cb1-28" data-line-number="28"></a>
<a class="sourceLine" id="cb1-29" data-line-number="29"><span class="co">#Format for PCA, calculate PCA, plot</span></a>
<a class="sourceLine" id="cb1-30" data-line-number="30">met_pr &lt;-<span class="st"> </span><span class="kw">pivot_wider</span>(metabolites_all, <span class="dt">names_from =</span> Compound, <span class="dt">values_from =</span> logValueZeros, <span class="dt">id_cols=</span> <span class="kw">c</span>(SampleID, study_site, diagnosis), <span class="dt">values_fill =</span> <span class="dv">0</span>)</a>
<a class="sourceLine" id="cb1-31" data-line-number="31">met_pca &lt;-<span class="st"> </span>met_pr <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">column_to_rownames</span>(<span class="dt">var =</span> <span class="st">&quot;SampleID&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="op">-</span>study_site, <span class="op">-</span>diagnosis) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.matrix</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">prcomp</span>()</a>
<a class="sourceLine" id="cb1-32" data-line-number="32">met_summary &lt;-<span class="st"> </span><span class="kw">as_tibble</span>(met_pca<span class="op">$</span>x, <span class="dt">rownames =</span> <span class="ot">NA</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rownames_to_column</span>(<span class="dt">var =</span> <span class="st">&quot;SampleID&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(SampleID, PC1, PC2, PC3, PC4) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">left_join</span>(<span class="kw">distinct</span>(<span class="kw">select</span>(metabolites_all, SampleID, diagnosis, study_site)))</a>
<a class="sourceLine" id="cb1-33" data-line-number="33">met_plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(met_summary, <span class="kw">aes</span>(<span class="dt">x =</span> PC1, <span class="dt">y =</span> PC2, <span class="dt">color =</span> diagnosis, <span class="dt">shape =</span> study_site)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="fl">1.5</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;Metabolite abundances&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>()</a>
<a class="sourceLine" id="cb1-34" data-line-number="34">met_plot</a></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-1-1.png" width="1056" /></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">met_vars &lt;-<span class="st"> </span>met_pca <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">tidy</span>(<span class="dt">matrix =</span> <span class="st">&quot;pcs&quot;</span>)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="co"># Remove large datasets</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="kw">rm</span>(metabolites_all, metabolites_cinc, metabolites_mgh)</a>
<a class="sourceLine" id="cb2-5" data-line-number="5"></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co">### Now do Taxa abundances</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7">taxa_cinc &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/taxa_profile_Cincinnati.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8"><span class="st">  </span><span class="kw">pivot_longer</span>(<span class="dt">names_to =</span> <span class="st">&quot;SampleID&quot;</span>, <span class="dt">cols =</span> <span class="op">-</span>Feature) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-9" data-line-number="9"><span class="st">  </span><span class="kw">left_join</span>(<span class="kw">select</span>(metadata_cinc, site_name, SampleID, SubjectID, diagnosis), <span class="dt">by =</span> <span class="st">&quot;SampleID&quot;</span>) <span class="co"># %&gt;% mutate(value_zeros = ifelse(is.na(value), 0, value))</span></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"></a>
<a class="sourceLine" id="cb2-11" data-line-number="11">taxa_mgh &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/taxa_profile_MGH.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="st">  </span><span class="kw">pivot_longer</span>(<span class="dt">names_to =</span> <span class="st">&quot;SampleID&quot;</span>, <span class="dt">cols =</span> <span class="op">-</span>Feature) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13"><span class="st">  </span><span class="kw">left_join</span>(<span class="kw">select</span>(metadata_mgh, site_name, SampleID, SubjectID, diagnosis), <span class="dt">by =</span> <span class="st">&quot;SampleID&quot;</span>) <span class="co">#%&gt;% </span></a>
<a class="sourceLine" id="cb2-14" data-line-number="14">  <span class="co">#mutate(value_zeros = ifelse(is.na(value), 0, value))</span></a>
<a class="sourceLine" id="cb2-15" data-line-number="15">taxa_all &lt;-<span class="st"> </span><span class="kw">rbind</span>(taxa_cinc, taxa_mgh)</a>
<a class="sourceLine" id="cb2-16" data-line-number="16"></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co">#Get species-level abundances, calculate Bray-Curtis dissimilarities between samples</span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18">bray_dist &lt;-<span class="st"> </span>taxa_all <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="op">-</span>diagnosis,<span class="op">-</span>SubjectID, <span class="op">-</span>site_name) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-19" data-line-number="19"><span class="st">  </span><span class="kw">filter</span>(<span class="kw">grepl</span>(<span class="st">&quot;s__.*&quot;</span>, Feature)<span class="op">|</span>Feature<span class="op">==</span><span class="st">&quot;UNKNOWN&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-20" data-line-number="20"><span class="st">  </span><span class="kw">pivot_wider</span>(<span class="dt">names_from =</span> <span class="st">&quot;Feature&quot;</span>, <span class="dt">values_from =</span> <span class="st">&quot;value&quot;</span>, <span class="dt">values_fill =</span> <span class="dv">0</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="st">  </span><span class="kw">column_to_rownames</span>(<span class="st">&quot;SampleID&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22"><span class="st">  </span><span class="kw">as.matrix</span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-23" data-line-number="23"><span class="st">  </span><span class="kw">vegdist</span>(<span class="dt">method =</span> <span class="st">&quot;bray&quot;</span>)</a>
<a class="sourceLine" id="cb2-24" data-line-number="24"></a>
<a class="sourceLine" id="cb2-25" data-line-number="25"><span class="co">#Perform principal coordinates analysis</span></a>
<a class="sourceLine" id="cb2-26" data-line-number="26">pcoa1 &lt;-<span class="st"> </span><span class="kw">pcoa</span>(bray_dist)</a>
<a class="sourceLine" id="cb2-27" data-line-number="27">pcoa_dat &lt;-<span class="st"> </span>pcoa1<span class="op">$</span>vectors <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>(<span class="dt">rownames =</span> <span class="ot">NA</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;SampleID&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">left_join</span>(<span class="kw">distinct</span>(<span class="kw">select</span>(taxa_all, SampleID, SubjectID, diagnosis, site_name)))</a>
<a class="sourceLine" id="cb2-28" data-line-number="28"></a>
<a class="sourceLine" id="cb2-29" data-line-number="29"><span class="co">#Visualize PCoA</span></a>
<a class="sourceLine" id="cb2-30" data-line-number="30">species_plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(pcoa_dat, <span class="kw">aes</span>(<span class="dt">x =</span> Axis<span class="fl">.1</span>, <span class="dt">y =</span> Axis<span class="fl">.2</span>, <span class="dt">color =</span> diagnosis, <span class="dt">shape =</span> site_name)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="fl">1.5</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;Species abundances&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">theme_classic</span>()</a>
<a class="sourceLine" id="cb2-31" data-line-number="31">species_plot</a></code></pre></div>
<p><img src="index_files/figure-html/unnamed-chunk-1-2.png" width="1056" /></p>
</div>
<div id="using-reaction-databases-to-link-taxa-and-genes-with-metabolites" class="section level1">
<h1><span class="header-section-number">2</span> Using reaction databases to link taxa and genes with metabolites</h1>
<p>We’ll start with a relatively narrow question. Can IBD-shifted metabolites be linked to variation in relevant microbial genes and pathways? For this analysis we will focus on the metabolites that have been assigned a known compound identification based on library search. Our main goal will be to demonstrate on a small scale how to query databases to do this kind of reference-informed integrative analysis.</p>
<p>First, we’ll load packages and read in data. We will use a few different packages for this section: the <code>tidyverse</code> suite for data manipulation and visualization, the <code>tidymodels</code> package for formatting our differential abundance results, the <code>webchem</code> package for mapping between different types of metabolite IDs, the <code>KEGGREST</code> package for querying the KEGG database, and the <code>patchwork</code> package for laying out plots with multiple panels.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">library</span>(tidymodels)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw">library</span>(webchem)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw">library</span>(KEGGREST)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="kw">library</span>(patchwork)</a>
<a class="sourceLine" id="cb3-6" data-line-number="6"><span class="kw">theme_set</span>(<span class="kw">theme_classic</span>())</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"></a>
<a class="sourceLine" id="cb3-8" data-line-number="8">metadata &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/subject_metadata_Cincinnati.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(data_type <span class="op">==</span><span class="st"> &quot;metabolomics&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">diagnosis =</span> <span class="kw">factor</span>(diagnosis, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;nonIBD&quot;</span>, <span class="st">&quot;UC&quot;</span>, <span class="st">&quot;CD&quot;</span>)))</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">metabolites_cinc &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/metabolite_profile_Cincinnati.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="st">  </span><span class="kw">rename_with</span>( <span class="op">~</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;_</span><span class="ch">\\</span><span class="st">(.*&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">gsub</span>(<span class="st">&quot; |</span><span class="ch">\\</span><span class="st">/&quot;</span>, <span class="st">&quot;_&quot;</span>, .x))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="st">  </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">is.na</span>(Metabolite)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-12" data-line-number="12"><span class="st">  </span><span class="kw">pivot_longer</span>(<span class="dt">names_to =</span> <span class="st">&quot;SampleID&quot;</span>, <span class="dt">cols =</span> <span class="kw">contains</span>(<span class="st">&quot;HSM&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-13" data-line-number="13"><span class="st">  </span><span class="kw">left_join</span>(<span class="kw">select</span>(metadata, SampleID, diagnosis), <span class="dt">by =</span> <span class="st">&quot;SampleID&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb3-14" data-line-number="14"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">value_zeros =</span> <span class="kw">ifelse</span>(<span class="kw">is.na</span>(value), <span class="dv">0</span>, value))</a>
<a class="sourceLine" id="cb3-15" data-line-number="15"></a>
<a class="sourceLine" id="cb3-16" data-line-number="16"><span class="co">## Get some initial summary statistics for this dataset</span></a>
<a class="sourceLine" id="cb3-17" data-line-number="17">metabolites_cinc <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarize</span>(<span class="kw">min</span>(value_zeros), <span class="kw">mean</span>(value_zeros), <span class="kw">median</span>(value_zeros), <span class="kw">max</span>(value_zeros))</a>
<a class="sourceLine" id="cb3-18" data-line-number="18"></a>
<a class="sourceLine" id="cb3-19" data-line-number="19"><span class="kw">length</span>(<span class="kw">unique</span>(metabolites_cinc<span class="op">$</span>Metabolite))</a>
<a class="sourceLine" id="cb3-20" data-line-number="20"></a>
<a class="sourceLine" id="cb3-21" data-line-number="21"><span class="co">## Let&#39;s log transform these metabolite values</span></a>
<a class="sourceLine" id="cb3-22" data-line-number="22">metabolites_cinc &lt;-<span class="st"> </span>metabolites_cinc <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">logValue =</span> <span class="kw">log1p</span>(value_zeros))</a>
<a class="sourceLine" id="cb3-23" data-line-number="23">metabolites_cinc <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarize</span>(<span class="kw">min</span>(logValue), <span class="kw">mean</span>(logValue), <span class="kw">median</span>(logValue), <span class="kw">max</span>(logValue))</a></code></pre></div>
<p>Next, let’s run a differential abundance analysis to identify which of compounds are shifted in abundance in either CD or UC compared to controls. For simplicity, we will fit multiple-hypothesis-adjusted linear models of our log-transformed metabolite intensities.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co">## </span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">ibd_diff_abundance &lt;-<span class="st"> </span>metabolites_cinc <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="st">  </span><span class="kw">group_by</span>(Compound, Method, m_z, RT, HMDB, Metabolite) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="st">  </span><span class="kw">do</span>(<span class="kw">tidy</span>(<span class="kw">lm</span>(.<span class="op">$</span>logValue <span class="op">~</span><span class="st"> </span>.<span class="op">$</span>diagnosis))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="st">  </span><span class="kw">filter</span>(term <span class="op">!=</span><span class="st"> &quot;(Intercept)&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="st">  </span><span class="kw">pivot_wider</span>(<span class="dt">id_cols =</span> <span class="kw">c</span>(<span class="st">&quot;Compound&quot;</span>, <span class="st">&quot;Method&quot;</span>, <span class="st">&quot;m_z&quot;</span>, <span class="st">&quot;RT&quot;</span>, <span class="st">&quot;HMDB&quot;</span>, <span class="st">&quot;Metabolite&quot;</span>), <span class="dt">names_from =</span> <span class="st">&quot;term&quot;</span>, <span class="dt">values_from =</span> <span class="kw">c</span>(<span class="st">&quot;statistic&quot;</span>, <span class="st">&quot;p.value&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="st">  </span><span class="kw">rename_with</span>( <span class="op">~</span><span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;_.$diagnosis&quot;</span>, <span class="st">&quot;&quot;</span>, .x, <span class="dt">fixed =</span> T)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">AdjustedPValCD =</span> <span class="kw">p.adjust</span>(p.valueCD), <span class="dt">AdjustedPValueUC =</span> <span class="kw">p.adjust</span>(p.valueUC)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="st">  </span><span class="kw">arrange</span>(AdjustedPValCD)</a>
<a class="sourceLine" id="cb4-10" data-line-number="10"></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="kw">View</span>(ibd_diff_abundance)</a>
<a class="sourceLine" id="cb4-12" data-line-number="12"></a>
<a class="sourceLine" id="cb4-13" data-line-number="13">top_cd_mets &lt;-<span class="st"> </span><span class="kw">filter</span>(ibd_diff_abundance, AdjustedPValCD <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span>)</a>
<a class="sourceLine" id="cb4-14" data-line-number="14">top_cd_mets <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count</span>(statisticCD <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="co"># Most of these are increased in CD</span></a></code></pre></div>
<p>Let’s plot a few of these.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">metabolites_cinc <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(Metabolite <span class="op">%in%</span><span class="st"> </span><span class="kw">pull</span>(top_cd_mets[<span class="dv">1</span><span class="op">:</span><span class="dv">8</span>,], Metabolite)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> diagnosis, <span class="dt">y =</span> logValue)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_boxplot</span>(<span class="dt">outlier.shape =</span> <span class="ot">NA</span>) <span class="op">+</span><span class="st"> </span><span class="kw">geom_jitter</span>(<span class="dt">width =</span> <span class="fl">0.25</span>, <span class="dt">size =</span> <span class="fl">0.4</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span>Metabolite, <span class="dt">scales =</span> <span class="st">&quot;free_y&quot;</span>)</a></code></pre></div>
<p>We are going to link these compounds to relevant genes and taxa. To do so, we are going to convert the HMDB compound IDs to KEGG compound IDs, and then query the KEGG database for reactions linked to those compounds.</p>
<p>While the database query runs, think about the question: What hypotheses are we testing in this analysis?</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co">## Fix HMDB IDs</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">top_cd_mets &lt;-<span class="st"> </span>top_cd_mets <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">HMDB2 =</span> <span class="kw">gsub</span>(<span class="st">&quot;*&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">gsub</span>(<span class="st">&quot;HMDB&quot;</span>, <span class="st">&quot;HMDB00&quot;</span>, HMDB), <span class="dt">fixed =</span> T))</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="co">#This will take a minute</span></a>
<a class="sourceLine" id="cb6-4" data-line-number="4">top_cd_mets &lt;-<span class="st"> </span>top_cd_mets <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="st">  </span><span class="kw">head</span>(<span class="dt">n =</span> <span class="dv">50</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-6" data-line-number="6"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">KEGG =</span> <span class="kw">cts_convert</span>(<span class="dt">query =</span> <span class="kw">gsub</span>(<span class="st">&quot;*&quot;</span>, <span class="st">&quot;&quot;</span>, HMDB2, <span class="dt">fixed =</span> T), <span class="dt">from =</span> <span class="st">&quot;Human Metabolome Database&quot;</span>, <span class="dt">to =</span> <span class="st">&quot;KEGG&quot;</span>)) <span class="co">##%&gt;% ungroup() %&gt;% mutate(KEGG = map(Metabolite, function(x) { names(keggFind(database = &quot;compound&quot;, query = x))}))</span></a></code></pre></div>
<p>Next, we will take our vector of KEGG compounds and retrieve a set of EC number enzymes and reactions known to be linked to those compounds.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="co"># Get KEGG compounds</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">kegg_comps &lt;-<span class="st"> </span>top_cd_mets <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">is.na</span>(KEGG)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="st">  </span><span class="kw">unnest</span>(<span class="dt">cols =</span> <span class="kw">c</span>(KEGG))</a>
<a class="sourceLine" id="cb7-4" data-line-number="4"></a>
<a class="sourceLine" id="cb7-5" data-line-number="5">kegg_comps &lt;-<span class="st"> </span>kegg_comps <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-6" data-line-number="6"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">ECRxn =</span> <span class="kw">map</span>(KEGG, <span class="cf">function</span>(x) {</a>
<a class="sourceLine" id="cb7-7" data-line-number="7">    <span class="kw">keggLink</span>(<span class="dt">target =</span> <span class="st">&quot;enzyme&quot;</span>, <span class="dt">source =</span> x)})) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="st">  </span><span class="kw">unnest</span>(<span class="dt">cols =</span> <span class="kw">c</span>(ECRxn))</a>
<a class="sourceLine" id="cb7-9" data-line-number="9"></a>
<a class="sourceLine" id="cb7-10" data-line-number="10"><span class="kw">View</span>(kegg_comps)</a>
<a class="sourceLine" id="cb7-11" data-line-number="11"></a>
<a class="sourceLine" id="cb7-12" data-line-number="12">kegg_comps &lt;-<span class="st"> </span>kegg_comps <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">ECRxn =</span> <span class="kw">gsub</span>(<span class="st">&quot;ec:&quot;</span>, <span class="st">&quot;&quot;</span>, ECRxn))</a></code></pre></div>
<p>Our metagenome data has been processed to quantify gene abundances linked to EC numbers, so now we can import that data and see what’s happening with these genes of interest. This is a large file so for now, we are going to filter to just the EC numbers we’re already interested in.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">ecs_cinc &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/ECnumber_profile_Cincinnati.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">EC2 =</span> <span class="kw">gsub</span>(<span class="st">&quot;: .*&quot;</span>, <span class="st">&quot;&quot;</span>, EC)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3"><span class="st">  </span><span class="kw">filter</span>(EC2 <span class="op">%in%</span><span class="st"> </span>kegg_comps<span class="op">$</span>ECRxn) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-4" data-line-number="4"><span class="st">  </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">grepl</span>(<span class="st">&quot;:&quot;</span>, Taxon)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="st">  </span><span class="kw">pivot_longer</span>(<span class="dt">cols =</span> <span class="kw">contains</span>(<span class="st">&quot;SM&quot;</span>), <span class="dt">names_to =</span> <span class="st">&quot;SampleID&quot;</span>, <span class="dt">values_to =</span> <span class="st">&quot;GeneAbundance&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-6" data-line-number="6"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">GeneAbundance =</span> <span class="kw">as.numeric</span>(GeneAbundance)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb8-7" data-line-number="7"><span class="st">  </span><span class="kw">inner_join</span>(metadata)</a>
<a class="sourceLine" id="cb8-8" data-line-number="8"></a>
<a class="sourceLine" id="cb8-9" data-line-number="9"></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="kw">length</span>(<span class="kw">unique</span>(ecs_cinc<span class="op">$</span>EC2))</a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="kw">length</span>(<span class="kw">unique</span>(ecs_cinc<span class="op">$</span>Taxon))</a>
<a class="sourceLine" id="cb8-12" data-line-number="12"></a>
<a class="sourceLine" id="cb8-13" data-line-number="13">ecs_cinc &lt;-<span class="st"> </span>ecs_cinc <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">left_join</span>(<span class="kw">select</span>(<span class="kw">ungroup</span>(kegg_comps), Metabolite, statisticCD, p.valueCD, KEGG, ECRxn), <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;EC2&quot;</span> =<span class="st"> &quot;ECRxn&quot;</span>))</a></code></pre></div>
<p>Let’s plot one example compound and its linked reactions.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">agmatine_abund &lt;-<span class="st"> </span><span class="kw">filter</span>(metabolites_cinc, Metabolite <span class="op">==</span><span class="st"> &quot;agmatine&quot;</span>)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2">ecs_agm &lt;-<span class="st"> </span><span class="kw">filter</span>(ecs_cinc, Metabolite <span class="op">==</span><span class="st"> &quot;agmatine&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb9-3" data-line-number="3"><span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(GeneAbundance))</a>
<a class="sourceLine" id="cb9-4" data-line-number="4"></a>
<a class="sourceLine" id="cb9-5" data-line-number="5">gene_met_data &lt;-<span class="st"> </span><span class="kw">left_join</span>(ecs_agm, agmatine_abund)</a>
<a class="sourceLine" id="cb9-6" data-line-number="6"></a>
<a class="sourceLine" id="cb9-7" data-line-number="7">highest_taxa &lt;-<span class="st"> </span>ecs_agm <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(Taxon) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-8" data-line-number="8"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">tot =</span> <span class="kw">sum</span>(GeneAbundance)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="st">  </span><span class="kw">arrange</span>(<span class="kw">desc</span>(tot)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">head</span>(<span class="dt">n =</span> <span class="dv">10</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(Taxon)</a>
<a class="sourceLine" id="cb9-10" data-line-number="10">gene_met_data &lt;-<span class="st"> </span>gene_met_data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">Taxon2 =</span> <span class="kw">ifelse</span>(Taxon <span class="op">%in%</span><span class="st"> </span>highest_taxa, Taxon, <span class="st">&quot;Other&quot;</span>))</a>
<a class="sourceLine" id="cb9-11" data-line-number="11"></a>
<a class="sourceLine" id="cb9-12" data-line-number="12">samp_order &lt;-<span class="st"> </span>agmatine_abund <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(logValue) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(SampleID)</a>
<a class="sourceLine" id="cb9-13" data-line-number="13"></a>
<a class="sourceLine" id="cb9-14" data-line-number="14">gene_met_data &lt;-<span class="st"> </span>gene_met_data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">mutate</span>(<span class="dt">SampleID =</span> <span class="kw">factor</span>(SampleID, <span class="dt">levels =</span> samp_order))</a>
<a class="sourceLine" id="cb9-15" data-line-number="15">gene_plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(gene_met_data, <span class="kw">aes</span>(<span class="dt">x =</span> SampleID, <span class="dt">fill =</span> Taxon2, <span class="dt">y =</span> GeneAbundance)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_bar</span>(<span class="dt">stat =</span> <span class="st">&quot;identity&quot;</span>, <span class="dt">position =</span> <span class="st">&quot;stack&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">facet_grid</span>(EC<span class="op">~</span>diagnosis, <span class="dt">scales =</span> <span class="st">&quot;free_x&quot;</span>, <span class="dt">space =</span> <span class="st">&quot;free_x&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">scale_fill_viridis_d</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_blank</span>()) <span class="co">#+ geom_point(aes(y = value))</span></a>
<a class="sourceLine" id="cb9-16" data-line-number="16"></a>
<a class="sourceLine" id="cb9-17" data-line-number="17">met_plot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(gene_met_data, <span class="kw">aes</span>(<span class="dt">x =</span> SampleID, <span class="dt">y =</span> logValue)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">facet_grid</span>(Metabolite<span class="op">~</span>diagnosis, <span class="dt">scales =</span> <span class="st">&quot;free_x&quot;</span>, <span class="dt">space =</span> <span class="st">&quot;free_x&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_blank</span>()) </a>
<a class="sourceLine" id="cb9-18" data-line-number="18"></a>
<a class="sourceLine" id="cb9-19" data-line-number="19">gene_plot <span class="op">+</span><span class="st"> </span>met_plot <span class="op">+</span><span class="st"> </span><span class="kw">plot_layout</span>(<span class="dt">nrow =</span> <span class="dv">2</span>, <span class="dt">heights =</span> <span class="kw">c</span>(<span class="dv">3</span>,<span class="dv">1</span>))</a></code></pre></div>
<p>What other ideas do you have of how we could improve this visualization or otherwise present this data? How could we systematically test whether these gene abundances are associated with the metabolite in question?</p>
<p>Finally, let’s query for some additional information about the most abundant associated gene family.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">rxn_info &lt;-<span class="st"> </span><span class="kw">keggGet</span>(<span class="st">&quot;ec:4.1.1.19&quot;</span>)[[<span class="dv">1</span>]]</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">rxn_info<span class="op">$</span>PATHWAY </a>
<a class="sourceLine" id="cb10-3" data-line-number="3">rxn_info<span class="op">$</span>CLASS</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">rxn_info<span class="op">$</span>SYSNAME</a></code></pre></div>
<div id="other-notes-and-follow-up-questions" class="section level2">
<h2><span class="header-section-number">2.1</span> Other notes and follow-up questions</h2>
<ul>
<li><p>You can extend this type of analysis and visualization by installing the <code>pathview</code> package to visualize data on top of KEGG pathway maps.</p></li>
<li><p>You can do a slightly different and more systematic version of this analysis (What microbial genes and taxa appear to be responsible or responsive to linked metabolite variation?) using a tool I developed during my PhD, <a href="http://borenstein-lab.github.io/MIMOSA2shiny/">MIMOSA2</a>.</p></li>
<li><p>We could alternatively link metagenomic and metabolomic data at the level of KEGG or Metacyc pathways. What are some tradeoffs in analyzing this data at the gene vs pathway levels?</p></li>
<li><p>What are the limitations of using this type of integrative approach? How could we strengthen this analysis?</p></li>
</ul>
</div>
</div>
<div id="predictive-analysis" class="section level1">
<h1><span class="header-section-number">3</span> Predictive analysis</h1>
<p>An alternative (and complementary!) approach to multi-omic analysis is to instead treat our dataset as just a “bag of features”: where we don’t assume anything about any specific feature’s meaning but simply apply high-dimensional analysis techniques to describe data structure and identify features possibly linked to our outcomes of interest. We’ll explore this now, using a machine learning analysis to evaluate what features best predict disease status.</p>
<!-- We are going to do this analysis using some new tools compared with what I've used in the past, that fit better with the concepts and strategies for data management that you've learned from the tidyverse tools, so bear with me. And note that this might seem to be organized in a somewhat odd way but it's been designed for repeatedly optimizing and expanding the initial analysis that we're going to do.
-->
<div id="classification" class="section level2">
<h2><span class="header-section-number">3.1</span> Classification</h2>
<p>In this analysis, we will investigate the questions: - How well do metabolomics and/or metagenomic features predict IBD diagnosis? - Can we identify biomarkers that are signatures of diagnosis?</p>
<p>These are <em>classification</em> questions - to answer them, we want to obtain a predictive model that can accurately predict what category (e.g. control or IBD) any sample belongs to, on the basis of its high-dimensional data. Then we will inspect the models to evaluate which features in our large dataset were most informative for those predictions.</p>
<p>We will use a commonly used type of machine learning model called a random forest. “Random Forest” describes an algorithm that combines predictions from many decision trees, each constructed from randomly sampled subsets of the data. Averaged together, these make better predictions than a single tree.</p>
<p><img src="img/random-forest-algorithm.png" /></p>
<p>Machine learning workflows can get quite convoluted, in large part to limit the risk of overparameterization and overfitting to dataset idiosyncrasies. (Overfitting is very easy to do when you have thousands of possible predictors.) In this analysis, we’ll work through a relatively standard workflow, using the <code>tidymodels</code> suite of packages:</p>
<ol style="list-style-type: decimal">
<li><p>Define the model choice and how we will split up our data for <strong>training</strong> and <strong>testing</strong> the model.</p></li>
<li><p>Use <strong>cross-validation</strong> to select hyperparameters for the model and get an initial idea of model accuracy (more on this below).</p></li>
<li><p>Fit a final model and examine which features are most predictive.</p></li>
<li><p>Test the model’s predictive ability on a new set of data.</p></li>
</ol>
<p>For these models, we will use a 5-fold cross-validation scheme to select the best model hyperparameters, and then we will calculate a final testing accuracy in an independent dataset - in this case, the samples from another study site cohort. k-fold cross-validation works by fitting k different models, each trained on 80% of the training data and tested on the remaining 20%, and then calculating the final performance statistics across all 5 folds.</p>
<p><img src="img/grid_search_cross_validation.png" /></p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw">library</span>(tidymodels)</a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="kw">library</span>(vip)</a>
<a class="sourceLine" id="cb11-4" data-line-number="4"></a>
<a class="sourceLine" id="cb11-5" data-line-number="5">metadata &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/subject_metadata_Cincinnati.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(data_type <span class="op">==</span><span class="st"> &quot;metagenomics&quot;</span>)</a>
<a class="sourceLine" id="cb11-6" data-line-number="6"></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="kw">glimpse</span>(metadata)</a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="co"># Set up our taxonomic data as a sample x feature matrix</span></a>
<a class="sourceLine" id="cb11-9" data-line-number="9"><span class="co">## Set up 5 divisions of the Cincinnati samples</span></a>
<a class="sourceLine" id="cb11-10" data-line-number="10">taxa_cin &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/taxa_profile_Cincinnati.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pivot_longer</span>(<span class="dt">names_to =</span> <span class="st">&quot;SampleID&quot;</span>, <span class="dt">cols =</span> <span class="op">-</span>Feature) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pivot_wider</span>(<span class="dt">names_from =</span> <span class="st">&quot;Feature&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">left_join</span>(<span class="kw">select</span>(metadata, SampleID, diagnosis), <span class="dt">by =</span> <span class="st">&quot;SampleID&quot;</span>)</a>
<a class="sourceLine" id="cb11-11" data-line-number="11"><span class="kw">colnames</span>(taxa_cin) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;|&quot;</span>, <span class="st">&quot;_&quot;</span>, <span class="kw">colnames</span>(taxa_cin), <span class="dt">fixed =</span> T) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot;]&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="dt">fixed =</span> T) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot;[&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="dt">fixed =</span> T)</a>
<a class="sourceLine" id="cb11-12" data-line-number="12"></a>
<a class="sourceLine" id="cb11-13" data-line-number="13">taxa_cin &lt;-<span class="st"> </span><span class="kw">mutate</span>(taxa_cin, <span class="dt">diagnosis =</span> <span class="kw">factor</span>(<span class="kw">ifelse</span>(diagnosis <span class="op">==</span><span class="st"> &quot;nonIBD&quot;</span>, <span class="st">&quot;nonIBD&quot;</span>, <span class="st">&quot;IBD&quot;</span>), <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;IBD&quot;</span>, <span class="st">&quot;nonIBD&quot;</span>)))</a></code></pre></div>
<p>Step 1: Set up training data scheme and model specification</p>
<p>Our first random forest will learn a model to predict IBD diagnosis (to start, just IBD or control) on the basis of taxonomic abundances.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># Set up cv scheme  </span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">folds1 &lt;-<span class="st"> </span><span class="kw">vfold_cv</span>(taxa_cin <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="op">-</span>SampleID), <span class="dt">v =</span> <span class="dv">5</span>, <span class="dt">strata =</span> <span class="st">&quot;diagnosis&quot;</span>)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3"></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="co">## Define model</span></a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="co"># Can also set a variety of other parameters here</span></a>
<a class="sourceLine" id="cb12-6" data-line-number="6">rand_model &lt;-<span class="st"> </span><span class="kw">rand_forest</span>(<span class="dt">trees =</span> <span class="kw">tune</span>(), <span class="dt">mtry =</span> <span class="kw">tune</span>()) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="st">  </span><span class="kw">set_engine</span>(<span class="st">&quot;ranger&quot;</span>, <span class="dt">importance =</span> <span class="st">&quot;impurity&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb12-8" data-line-number="8"><span class="st">  </span><span class="kw">set_mode</span>(<span class="st">&quot;classification&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb12-9" data-line-number="9"><span class="st">  </span><span class="kw">translate</span>()</a>
<a class="sourceLine" id="cb12-10" data-line-number="10"></a>
<a class="sourceLine" id="cb12-11" data-line-number="11"><span class="co"># Grid of possible parameters for our model</span></a>
<a class="sourceLine" id="cb12-12" data-line-number="12">tree_grid &lt;-<span class="st"> </span><span class="kw">grid_regular</span>(<span class="kw">mtry</span>(<span class="dt">range =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">200</span>)), <span class="kw">trees</span>(<span class="dt">range =</span> <span class="kw">c</span>(<span class="dv">100</span>, <span class="dv">800</span>)), <span class="dt">levels =</span> <span class="dv">3</span>)</a>
<a class="sourceLine" id="cb12-13" data-line-number="13"></a>
<a class="sourceLine" id="cb12-14" data-line-number="14"><span class="co">## Our overall workflow will be fitting random forests to predict diagnosis</span></a>
<a class="sourceLine" id="cb12-15" data-line-number="15">rf_wf &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb12-16" data-line-number="16"><span class="st">  </span><span class="kw">workflow</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-17" data-line-number="17"><span class="st">  </span><span class="kw">add_model</span>(rand_model) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-18" data-line-number="18"><span class="st">  </span><span class="kw">add_formula</span>(diagnosis <span class="op">~</span><span class="st"> </span>.)</a></code></pre></div>
<p>Step 2: Use cross-validation to select hyperparameters for the model and get an initial idea of model quality</p>
<p>We will use cross-validation performance to choose parameters for our model. Random forests have 2 main parameters (there are also others that we will leave as the defaults): the number of decision trees to make up the whole model (<code>num.trees</code>), and the number of variables randomly selected for each tree (<code>mtry</code>). The optimal values for these depend on dataset characteristics. We will pick the values that perform the best from a selection.</p>
<p>We’ll evaluate our predictive models using 2 common metrics: - Accuracy: total correct predictions divided by total predictions or samples - Area under the ROC curve (AUC): This is a commonly used metric in classification analyses. The predictions produced by the model are actually probabilities: the estimated probability that that particular sample belongs to each class. An ROC curve accounts for the idea that we could use probability cutoffs to assign a sample to each class.</p>
<!-- #https://www.tidymodels.org/start/tuning/ -->
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="co">## This will take a little while</span></a>
<a class="sourceLine" id="cb13-2" data-line-number="2">rf_fit_cv &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb13-3" data-line-number="3"><span class="st">  </span>rf_wf <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-4" data-line-number="4"><span class="st">  </span><span class="kw">tune_grid</span>(<span class="dt">resamples =</span> folds1,</a>
<a class="sourceLine" id="cb13-5" data-line-number="5">            <span class="dt">grid =</span> tree_grid, <span class="dt">control =</span> <span class="kw">control_grid</span>(<span class="dt">save_pred =</span> T))</a>
<a class="sourceLine" id="cb13-6" data-line-number="6"></a>
<a class="sourceLine" id="cb13-7" data-line-number="7"><span class="co">## Let&#39;s look at the ROC curve from the predictions from each model</span></a>
<a class="sourceLine" id="cb13-8" data-line-number="8">cv_predictions &lt;-<span class="st"> </span><span class="kw">collect_predictions</span>(rf_fit_cv)</a>
<a class="sourceLine" id="cb13-9" data-line-number="9"></a>
<a class="sourceLine" id="cb13-10" data-line-number="10">cv_predictions <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">conf_mat</span>(<span class="dt">truth =</span> diagnosis, .pred_class)</a>
<a class="sourceLine" id="cb13-11" data-line-number="11"></a>
<a class="sourceLine" id="cb13-12" data-line-number="12">cv_predictions <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(.config <span class="op">==</span><span class="st"> &quot;Model1&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-13" data-line-number="13"><span class="st">  </span><span class="kw">roc_curve</span>(<span class="dt">truth =</span> diagnosis, .pred_IBD) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">autoplot</span>()</a>
<a class="sourceLine" id="cb13-14" data-line-number="14"></a>
<a class="sourceLine" id="cb13-15" data-line-number="15">cv_predictions <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(.config) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-16" data-line-number="16"><span class="st">  </span><span class="kw">roc_curve</span>(<span class="dt">truth =</span> diagnosis, .pred_IBD) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> <span class="dv">1</span> <span class="op">-</span><span class="st"> </span>specificity, <span class="dt">y =</span> sensitivity, <span class="dt">color =</span> .config)) <span class="op">+</span><span class="st"> </span><span class="kw">geom_line</span>()</a></code></pre></div>
<p>We’ll make a plot of performance across all of our different parameter combinations to choose what to use for our final model:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co"># CV accuracy with our different parameter combinations</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2"><span class="kw">collect_metrics</span>(rf_fit_cv) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pivot_wider</span>(<span class="dt">names_from =</span> <span class="st">&quot;.metric&quot;</span>, <span class="dt">values_from =</span> <span class="st">&quot;mean&quot;</span>, <span class="dt">id_cols =</span> <span class="kw">c</span>(<span class="st">&quot;mtry&quot;</span>, <span class="st">&quot;trees&quot;</span>, <span class="st">&quot;.config&quot;</span>))</a>
<a class="sourceLine" id="cb14-3" data-line-number="3"></a>
<a class="sourceLine" id="cb14-4" data-line-number="4">rf_fit_cv <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-5" data-line-number="5"><span class="st">  </span><span class="kw">collect_metrics</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-6" data-line-number="6"><span class="st">  </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> trees, <span class="dt">y =</span> mean, <span class="dt">color =</span> <span class="kw">factor</span>(mtry))) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-7" data-line-number="7"><span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">size =</span> <span class="fl">1.5</span>, <span class="dt">alpha =</span> <span class="fl">0.6</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-8" data-line-number="8"><span class="st">  </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-9" data-line-number="9"><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="op">~</span><span class="st"> </span>.metric, <span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>, <span class="dt">nrow =</span> <span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-10" data-line-number="10"><span class="st">  </span><span class="kw">scale_x_log10</span>(<span class="dt">labels =</span> scales<span class="op">::</span><span class="kw">label_number</span>()) <span class="op">+</span></a>
<a class="sourceLine" id="cb14-11" data-line-number="11"><span class="st">  </span><span class="kw">scale_color_viridis_d</span>(<span class="dt">option =</span> <span class="st">&quot;plasma&quot;</span>, <span class="dt">begin =</span> <span class="fl">.9</span>, <span class="dt">end =</span> <span class="dv">0</span>)</a>
<a class="sourceLine" id="cb14-12" data-line-number="12"></a>
<a class="sourceLine" id="cb14-13" data-line-number="13">rf_fit_cv <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">show_best</span>(<span class="st">&quot;roc_auc&quot;</span>) </a>
<a class="sourceLine" id="cb14-14" data-line-number="14">best_rf &lt;-<span class="st"> </span>rf_fit_cv <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb14-15" data-line-number="15"><span class="st">  </span><span class="kw">select_by_one_std_err</span>(<span class="dt">metric =</span> <span class="st">&quot;roc_auc&quot;</span>, mtry, trees)</a></code></pre></div>
<p>Step 3: Fit final model and examine which features are most predictive</p>
<p>Now that we’ve selected reasonable parameters, we can fit our final training model and examine which features appear to be most informative for predicting IBD status in this dataset.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="co">#Tell the workflow to only use the best model parameters now</span></a>
<a class="sourceLine" id="cb15-2" data-line-number="2">final_wf &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb15-3" data-line-number="3"><span class="st">  </span>rf_wf <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb15-4" data-line-number="4"><span class="st">  </span><span class="kw">finalize_workflow</span>(best_rf)</a>
<a class="sourceLine" id="cb15-5" data-line-number="5"></a>
<a class="sourceLine" id="cb15-6" data-line-number="6"><span class="co"># Fit the model to the whole dataset</span></a>
<a class="sourceLine" id="cb15-7" data-line-number="7">final_rf &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb15-8" data-line-number="8"><span class="st">  </span>final_wf <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb15-9" data-line-number="9"><span class="st">  </span><span class="kw">fit</span>(<span class="dt">data =</span> taxa_cin) </a>
<a class="sourceLine" id="cb15-10" data-line-number="10"></a>
<a class="sourceLine" id="cb15-11" data-line-number="11"><span class="co">#Look at the final training AUC values</span></a>
<a class="sourceLine" id="cb15-12" data-line-number="12">final_pred &lt;-<span class="st"> </span><span class="kw">bind_cols</span>(<span class="kw">select</span>(taxa_cin, diagnosis), </a>
<a class="sourceLine" id="cb15-13" data-line-number="13">                        <span class="kw">predict</span>(final_rf, taxa_cin, <span class="dt">type =</span> <span class="st">&quot;class&quot;</span>), </a>
<a class="sourceLine" id="cb15-14" data-line-number="14">                        <span class="kw">predict</span>(final_rf, taxa_cin, <span class="dt">type =</span> <span class="st">&quot;prob&quot;</span>))</a>
<a class="sourceLine" id="cb15-15" data-line-number="15">final_pred <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">conf_mat</span>(<span class="dt">truth =</span> diagnosis, .pred_class) </a>
<a class="sourceLine" id="cb15-16" data-line-number="16"></a>
<a class="sourceLine" id="cb15-17" data-line-number="17">final_pred <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb15-18" data-line-number="18"><span class="st">  </span><span class="kw">roc_curve</span>(<span class="dt">truth =</span> diagnosis, .pred_IBD) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">autoplot</span>()</a>
<a class="sourceLine" id="cb15-19" data-line-number="19"></a>
<a class="sourceLine" id="cb15-20" data-line-number="20"><span class="kw">roc_auc</span>(final_pred, <span class="dt">truth =</span> diagnosis, .pred_IBD) </a></code></pre></div>
<p>Looks like a pretty strong model! We can also look at which particular taxa tended to be most informative of IBD status in this final model. This is quantified here using a metric called Gini impurity.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">final_rf <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="st">  </span><span class="kw">pull_workflow_fit</span>() <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb16-3" data-line-number="3"><span class="st">  </span><span class="kw">vip</span>(<span class="dt">num_features =</span> <span class="dv">15</span>)</a></code></pre></div>
<p>Step 4: Test final model on a new set of data</p>
<p>Even though our model seems to perform really well, this is almost always true on a training dataset (the model is optimized for that after all). To really get a sense of whether these particular combinations of taxa are informative for predicting IBD status, we need to apply our model to a totally new dataset. In this case, we will try predicting IBD status with the same model (same parameters, same final fit decision trees) on a new, independent dataset - the samples from the MGH study site.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">metadata_mgh &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/subject_metadata_MGH.tsv&quot;</span>)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">taxa_mgh &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/taxa_profile_MGH.tsv&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pivot_longer</span>(<span class="dt">names_to =</span> <span class="st">&quot;SampleID&quot;</span>, <span class="dt">cols =</span> <span class="op">-</span>Feature) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pivot_wider</span>(<span class="dt">names_from =</span> <span class="st">&quot;Feature&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">left_join</span>(<span class="kw">select</span>(metadata_mgh, SampleID, diagnosis), <span class="dt">by =</span> <span class="st">&quot;SampleID&quot;</span>)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="kw">colnames</span>(taxa_mgh) &lt;-<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;|&quot;</span>, <span class="st">&quot;_&quot;</span>, <span class="kw">colnames</span>(taxa_mgh), <span class="dt">fixed =</span> T) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot;]&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="dt">fixed =</span> T) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">gsub</span>(<span class="dt">pattern =</span> <span class="st">&quot;[&quot;</span>, <span class="dt">replacement =</span> <span class="st">&quot;&quot;</span>, <span class="dt">fixed =</span> T)</a>
<a class="sourceLine" id="cb17-4" data-line-number="4"></a>
<a class="sourceLine" id="cb17-5" data-line-number="5">taxa_mgh &lt;-<span class="st"> </span><span class="kw">mutate</span>(taxa_mgh, <span class="dt">diagnosis =</span> <span class="kw">factor</span>(<span class="kw">ifelse</span>(diagnosis <span class="op">==</span><span class="st"> &quot;nonIBD&quot;</span>, <span class="st">&quot;nonIBD&quot;</span>, <span class="st">&quot;IBD&quot;</span>), <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;IBD&quot;</span>, <span class="st">&quot;nonIBD&quot;</span>)))</a>
<a class="sourceLine" id="cb17-6" data-line-number="6"></a>
<a class="sourceLine" id="cb17-7" data-line-number="7">taxa_mgh &lt;-<span class="st"> </span>taxa_mgh <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">bind_cols</span>(</a>
<a class="sourceLine" id="cb17-8" data-line-number="8">  <span class="kw">predict</span>(final_rf, <span class="dt">new_data =</span> taxa_mgh),</a>
<a class="sourceLine" id="cb17-9" data-line-number="9">  <span class="kw">predict</span>(final_rf, <span class="dt">new_data =</span> taxa_mgh, <span class="dt">type =</span> <span class="st">&quot;prob&quot;</span>))</a>
<a class="sourceLine" id="cb17-10" data-line-number="10"></a>
<a class="sourceLine" id="cb17-11" data-line-number="11"><span class="kw">roc_curve</span>(taxa_mgh, <span class="dt">truth =</span> diagnosis, .pred_IBD) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">autoplot</span>()</a>
<a class="sourceLine" id="cb17-12" data-line-number="12"><span class="kw">roc_auc</span>(taxa_mgh, <span class="dt">truth =</span> diagnosis, .pred_IBD)</a>
<a class="sourceLine" id="cb17-13" data-line-number="13"></a>
<a class="sourceLine" id="cb17-14" data-line-number="14">taxa_mgh <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">conf_mat</span>(<span class="dt">truth =</span> diagnosis, .pred_class)</a>
<a class="sourceLine" id="cb17-15" data-line-number="15">taxa_mgh <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarize</span>(<span class="dt">Accuracy =</span> <span class="kw">sum</span>(diagnosis <span class="op">==</span><span class="st"> </span>.pred_class)<span class="op">/</span><span class="kw">length</span>(diagnosis))</a></code></pre></div>
</div>
<div id="follow-up-questions" class="section level2">
<h2><span class="header-section-number">3.2</span> Follow up questions</h2>
<ul>
<li><p>Try this with another data type: Does the metabolite data predict disease subtype more accurately than the taxonomic data? What about both datasets together, and/or some of the additional metadata such as sex and age?</p></li>
<li><p>Try another formulation of the problem: what about a multi-class classifier to distinguish UC, CD, and nonIBD?</p></li>
<li><p>How well do you think a model would perform based solely on the 10 or 15 most important taxa?</p></li>
<li><p>Do you think this training-testing setup was a good approach to prevent overfitting and over-interpretation? How could we refine this strategy?</p></li>
<li><p>Do you think the random forest algorithm is a good choice of classification tool for this dataset? Why or why not? Look into support vector machines and elastic net logistic regression as alternative approaches.</p></li>
</ul>
<!--
# Extensions

These are all optional suggestions for ways to get more practice with R and omics analysis on this dataset. 

- Perform a descriptive analysis of the taxonomic, gene family, and/or metabolite data using the tools you applied in the amplicon sequencing tutorial:
- Principal coordinates analysis based on Bray-Curtis dissimilarities between samples
- PERMANOVA test to assess whether dissimilarities can be explained by disease status or other metadata variables (e.g. age, study site)
- If you generate ordination plots for two different data types, you can perform a Procrustes analysis to compare them. This analysis evaluates whether 2 different dimensional reductions (typically based on 2 different datasets from the same samples) find a similar distribution fo those samples. Look up the `procrustes()` function in the R package `vegan` to learn more.
- Perform a differential abundance analysis of metagenomic species abundances using ALDeX2

-->
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
