<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Multi-omics: Pre-Class Setup</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/spacelab.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { font-weight: bold; } /* Alert */
code span.an { font-style: italic; } /* Annotation */
code span.cf { font-weight: bold; } /* ControlFlow */
code span.co { font-style: italic; } /* Comment */
code span.cv { font-style: italic; } /* CommentVar */
code span.do { font-style: italic; } /* Documentation */
code span.dt { text-decoration: underline; } /* DataType */
code span.er { font-weight: bold; } /* Error */
code span.in { font-style: italic; } /* Information */
code span.kw { font-weight: bold; } /* Keyword */
code span.pp { font-weight: bold; } /* Preprocessor */
code span.wa { font-style: italic; } /* Warning */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 52px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h2 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h3 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h4 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h5 {
  padding-top: 57px;
  margin-top: -57px;
}
.section h6 {
  padding-top: 57px;
  margin-top: -57px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Multi-omics analysis tutorial, 2021</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="workshop_prep.html">Advance setup</a>
</li>
<li>
  <a href="index.html">Tutorial</a>
</li>
<li>
  <a href="full_output.html">Tutorial with full code and output</a>
</li>
<li>
  <a href="slides.pdf">Tutorial slides</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Multi-omics: Pre-Class Setup</h1>

</div>


<p>If you run into any setup problems or questions, please feel free to email me at cecilia.noecker at ucsf.edu.</p>
<div id="set-up-your-r-project-and-install-packages" class="section level1">
<h1><span class="header-section-number">1</span> Set up your R project and install packages</h1>
<p>Hopefully you have already installed R, RStudio, and the tidyverse suite of packages for the previous amplicon sequencing tutorial. If you still need to install these, you can find instructions for doing so <a href="https://datacarpentry.org/R-ecology-lesson/#Preparations">here</a>. The linked Data/Software Carpentry website includes a lot of other useful resources for learning R and doing scientific data analysis.</p>
<p>For this session, we will create an R Project to organize our analyses. To do so, open RStudio, select File -&gt; New Project… and create a new folder for this project (name it whatever you want).</p>
<p>You will also need to install the following packages:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw">install.packages</span>(<span class="st">&quot;tidymodels&quot;</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw">install.packages</span>(<span class="st">&quot;vip&quot;</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"><span class="kw">install.packages</span>(<span class="st">&quot;webchem&quot;</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4"><span class="kw">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">&quot;KEGGREST&quot;</span>)</a></code></pre></div>
</div>
<div id="download-data" class="section level1">
<h1><span class="header-section-number">2</span> Download data</h1>
<p>In this tutorial, we’ll analyze processed datasets made available as part of the large-scale Integrative Human Microbiome Project data resource (a.k.a. HMP2). The study and data are described in <a href="https://www.nature.com/articles/s41564-018-0306-4">Franzosa et al Nature Microbiology, 2019</a>.</p>
<p>In this study, the researchers profiled stool samples from a cohort of individuals with ulcerative colitis, Crohn’s disease, and non-IBD controls. Most subjects were sampled at several time points. These individuals were recruited at 5 different study sites. In the interest of computational feasibility we will focus our analysis on a subset of samples from just 2 sites. Stool samples collected from all subjects were profiled using metagenomics and metabolomics assays, as shown in Figure 1a from the publication:</p>
<p><img src="img/fig1.png" /></p>
<p><strong>Note:</strong> Two options for downloading the data we’ll use in the workshop are described below. If your computer has at least 8GB of memory and ~500MB of storage available, I recommend working through Option 1, which describes how to download the full dataset from the HMP portal and work through some data processing to prepare it for analysis. This will allow you to practice your R skills and see an example workflow of setting up a re-analysis of a public dataset. However, if your computer has less memory, it might have a hard time processing these tables, in which case you can follow Option 2 to directly download smaller pre-processed datasets.</p>
<div id="option-1-download-and-process-directly-from-the-hmp-portal" class="section level2">
<h2><span class="header-section-number">2.1</span> Option 1: Download and process directly from the HMP portal</h2>
<p>Follow these instructions to download each of the various data files:</p>
<ul>
<li><strong>Metadata:</strong> From <a href="https://ibdmdb.org/tunnel/public/summary.html">this link</a>, select “Download HMP2 Metadata”.</li>
<li><strong>Metagenomic data:</strong> At <a href="https://ibdmdb.org/tunnel/public/HMP2/WGS/1818/products">this link</a>, select “Merged tables”, then click to download 3 files: <code>taxonomic_profiles_3.tsv.gz</code>, <code>path_abundances_3.tsv.gz</code>, and <code>ecs_3.tsv.gz</code>. These files are taxonomic abundances, pathway abundances, and abundances of genes linked to EC numbers, produced by the Metaphlan3 + Humann3 pipeline. You can read more about those tools <a href="https://elifesciences.org/articles/65088">here</a>.</li>
<li><strong>Metabolomics data:</strong> From <a href="https://ibdmdb.org/tunnel/public/HMP2/Metabolites/1723/products">this page</a>, click to download the file <code>HMP2_metabolomics.csv.gz</code></li>
</ul>
<!-- - Finally, you will also need to download [this table of metabolite ID mappings](). -->
<div id="read-in-data-process-and-filter" class="section level3">
<h3><span class="header-section-number">2.1.1</span> Read in data, process and filter</h3>
<p>The code below goes through to HMP2 files to process the data as follows: 1) clean up column names and table organization, 2) subset to a smaller set of samples from 2 study sites, 3) filter low-prevalence features that only appear in a small number of samples.</p>
<div id="metadata" class="section level4">
<h4><span class="header-section-number">2.1.1.1</span> Metadata</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co"># Make a sub-folder for storing data</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw">dir.create</span>(<span class="st">&quot;data/&quot;</span>)</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="co"># Read in metadata</span></a>
<a class="sourceLine" id="cb2-6" data-line-number="6">metadata &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;data/hmp2_metadata.csv&quot;</span>, <span class="dt">col_types =</span> <span class="kw">cols</span>(<span class="dt">.default =</span> <span class="st">&quot;c&quot;</span>))</a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="co"># How many samples of each type?</span></a>
<a class="sourceLine" id="cb2-8" data-line-number="8">metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count</span>(data_type)</a>
<a class="sourceLine" id="cb2-9" data-line-number="9"></a>
<a class="sourceLine" id="cb2-10" data-line-number="10"><span class="co">## Keep just the metadata we want, rename columns we will use frequently</span></a>
<a class="sourceLine" id="cb2-11" data-line-number="11">metadata &lt;-<span class="st"> </span>metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(data_type <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;metabolomics&quot;</span>, <span class="st">&quot;metagenomics&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">rename</span>(<span class="dt">SampleID =</span> <span class="st">`</span><span class="dt">External ID</span><span class="st">`</span>, <span class="dt">SubjectID =</span> <span class="st">`</span><span class="dt">Participant ID</span><span class="st">`</span>)</a>
<a class="sourceLine" id="cb2-12" data-line-number="12"><span class="co">#Only keep subjects with both metabolomics and metagenomics samples</span></a>
<a class="sourceLine" id="cb2-13" data-line-number="13">subjects_keep &lt;-<span class="st"> </span>metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count</span>(SubjectID, data_type, <span class="dt">.drop =</span> F) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count</span>(SubjectID) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(n<span class="op">==</span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb2-14" data-line-number="14">metadata &lt;-<span class="st"> </span>metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(SubjectID <span class="op">%in%</span><span class="st"> </span>subjects_keep<span class="op">$</span>SubjectID)</a>
<a class="sourceLine" id="cb2-15" data-line-number="15">metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count</span>(site_name, diagnosis, data_type) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">print</span>(<span class="dt">n=</span><span class="dv">30</span>)</a>
<a class="sourceLine" id="cb2-16" data-line-number="16"></a>
<a class="sourceLine" id="cb2-17" data-line-number="17"><span class="co">#Remove sample duplicates that were resequenced</span></a>
<a class="sourceLine" id="cb2-18" data-line-number="18">bad_metagenomics_samps &lt;-<span class="st"> </span>metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(data_type <span class="op">==</span><span class="st"> &quot;metagenomics&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(<span class="dt">NewID =</span> <span class="kw">gsub</span>(<span class="st">&quot;_P$&quot;</span>, <span class="st">&quot;&quot;</span>, SampleID)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(n <span class="op">==</span><span class="st"> </span><span class="dv">2</span>)</a>
<a class="sourceLine" id="cb2-19" data-line-number="19">metadata &lt;-<span class="st"> </span>metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(<span class="op">!</span>SampleID <span class="op">%in%</span><span class="st"> </span>bad_metagenomics_samps<span class="op">$</span>NewID)  </a>
<a class="sourceLine" id="cb2-20" data-line-number="20"></a>
<a class="sourceLine" id="cb2-21" data-line-number="21"><span class="co">## We will just look at the samples from the Cincinnati site to start, and will use the data from the MGH dataset as a test dataset for the machine learning analysis </span></a>
<a class="sourceLine" id="cb2-22" data-line-number="22">metadata_cinc_mgh &lt;-<span class="st"> </span>metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(site_name <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Cincinnati&quot;</span>, <span class="st">&quot;MGH&quot;</span>))</a>
<a class="sourceLine" id="cb2-23" data-line-number="23"></a>
<a class="sourceLine" id="cb2-24" data-line-number="24"><span class="co"># Only save metadata variables at the participant level (not the sample level), and that actually differ between participants</span></a>
<a class="sourceLine" id="cb2-25" data-line-number="25">metadata_long &lt;-<span class="st"> </span><span class="kw">pivot_longer</span>(metadata, <span class="dt">cols =</span> <span class="op">!</span>SampleID <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span>SubjectID, <span class="dt">names_to =</span> <span class="st">&quot;name&quot;</span>, <span class="dt">values_to =</span> <span class="st">&quot;value&quot;</span>, <span class="dt">values_transform =</span> <span class="kw">list</span>(<span class="dt">value =</span> as.character))</a>
<a class="sourceLine" id="cb2-26" data-line-number="26">subject_level_vars &lt;-<span class="st"> </span>metadata_long <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(name, SubjectID) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-27" data-line-number="27"><span class="st">  </span><span class="kw">summarise</span>(<span class="dt">numVals =</span> <span class="kw">length</span>(<span class="kw">unique</span>(value[<span class="op">!</span><span class="kw">is.na</span>(value)]))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-28" data-line-number="28"><span class="st">  </span><span class="kw">group_by</span>(name) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="dt">fracSubs =</span> <span class="kw">length</span>(numVals[numVals <span class="op">&lt;</span><span class="st"> </span><span class="dv">2</span>])<span class="op">/</span><span class="kw">length</span>(numVals)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb2-29" data-line-number="29"><span class="st">  </span><span class="kw">filter</span>(fracSubs <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(name) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unique</span>()</a>
<a class="sourceLine" id="cb2-30" data-line-number="30"></a>
<a class="sourceLine" id="cb2-31" data-line-number="31">tot_vals &lt;-<span class="st"> </span>metadata_long <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(SampleID <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(metadata_cinc_mgh<span class="op">$</span>SampleID) <span class="op">&amp;</span><span class="st"> </span>name <span class="op">%in%</span><span class="st"> </span>subject_level_vars) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(name) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">summarise</span>(<span class="dt">numVals =</span> <span class="kw">length</span>(<span class="kw">unique</span>(value[<span class="op">!</span><span class="kw">is.na</span>(value)]))) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(numVals <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>)</a>
<a class="sourceLine" id="cb2-32" data-line-number="32"></a>
<a class="sourceLine" id="cb2-33" data-line-number="33">subject_metadata &lt;-<span class="st"> </span>metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="kw">any_of</span>(<span class="kw">c</span>(<span class="st">&quot;SubjectID&quot;</span>, <span class="st">&quot;SampleID&quot;</span>, <span class="st">&quot;data_type&quot;</span>, tot_vals<span class="op">$</span>name)))</a>
<a class="sourceLine" id="cb2-34" data-line-number="34">subject_metadata_mgh_cinc &lt;-<span class="st"> </span>subject_metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(site_name <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;MGH&quot;</span>, <span class="st">&quot;Cincinnati&quot;</span>))</a>
<a class="sourceLine" id="cb2-35" data-line-number="35"></a>
<a class="sourceLine" id="cb2-36" data-line-number="36"><span class="co"># Save read QC information for metagenomic samples</span></a>
<a class="sourceLine" id="cb2-37" data-line-number="37">read_metadata_mgh_cinc &lt;-<span class="st"> </span>metadata <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(data_type <span class="op">==</span><span class="st"> &quot;metagenomics&quot;</span> <span class="op">&amp;</span><span class="st"> </span>site_name <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;MGH&quot;</span>, <span class="st">&quot;Cincinnati&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="kw">contains</span>(<span class="st">&quot;reads&quot;</span>)<span class="op">|</span><span class="kw">contains</span>(<span class="st">&quot;External ID&quot;</span>)<span class="op">|</span><span class="kw">contains</span>(<span class="st">&quot;site_name&quot;</span>))</a>
<a class="sourceLine" id="cb2-38" data-line-number="38"></a>
<a class="sourceLine" id="cb2-39" data-line-number="39"><span class="co"># Save cleaned-up tables from each site</span></a>
<a class="sourceLine" id="cb2-40" data-line-number="40"><span class="cf">for</span>(site <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;MGH&quot;</span>, <span class="st">&quot;Cincinnati&quot;</span>)){</a>
<a class="sourceLine" id="cb2-41" data-line-number="41">  <span class="kw">write_tsv</span>(<span class="kw">filter</span>(subject_metadata_mgh_cinc, site_name <span class="op">==</span><span class="st"> </span>site), <span class="dt">path =</span> <span class="kw">paste0</span>(<span class="st">&quot;data/subject_metadata_&quot;</span>, site, <span class="st">&quot;.tsv&quot;</span>))</a>
<a class="sourceLine" id="cb2-42" data-line-number="42">  <span class="kw">write_tsv</span>(<span class="kw">filter</span>(read_metadata_mgh_cinc, site_name <span class="op">==</span><span class="st"> </span>site), <span class="dt">path =</span> <span class="kw">paste0</span>(<span class="st">&quot;data/read_metadata_&quot;</span>, site, <span class="st">&quot;.tsv&quot;</span>))</a>
<a class="sourceLine" id="cb2-43" data-line-number="43">}</a></code></pre></div>
</div>
<div id="taxonomic-data" class="section level4">
<h4><span class="header-section-number">2.1.1.2</span> Taxonomic data</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># Read in file, get sample subsets, rename columns, save</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"></a>
<a class="sourceLine" id="cb3-3" data-line-number="3">taxa_abundances &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/taxonomic_profiles_3.tsv.gz&quot;</span>)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="cf">for</span>(site <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;MGH&quot;</span>, <span class="st">&quot;Cincinnati&quot;</span>)){</a>
<a class="sourceLine" id="cb3-5" data-line-number="5">  site_samples &lt;-<span class="st"> </span><span class="kw">filter</span>(subject_metadata_mgh_cinc, site_name <span class="op">==</span><span class="st"> </span>site)<span class="op">$</span>SampleID</a>
<a class="sourceLine" id="cb3-6" data-line-number="6">  taxa_abund &lt;-<span class="st"> </span>taxa_abundances <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="kw">any_of</span>(<span class="kw">c</span>(<span class="kw">names</span>(taxa_abundances)[<span class="dv">1</span>], <span class="kw">paste0</span>(site_samples, <span class="st">&quot;_profile&quot;</span>))))</a>
<a class="sourceLine" id="cb3-7" data-line-number="7">  <span class="kw">colnames</span>(taxa_abund) =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Feature&quot;</span>, <span class="kw">gsub</span>(<span class="st">&quot;_profile&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">colnames</span>(taxa_abund)[<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(taxa_abund)]))</a>
<a class="sourceLine" id="cb3-8" data-line-number="8">  <span class="kw">write_tsv</span>(taxa_abund, <span class="dt">path =</span> <span class="kw">paste0</span>(<span class="st">&quot;data/taxa_profile_&quot;</span>, site, <span class="st">&quot;.tsv&quot;</span>))</a>
<a class="sourceLine" id="cb3-9" data-line-number="9">}</a>
<a class="sourceLine" id="cb3-10" data-line-number="10"><span class="kw">rm</span>(taxa_abundances)</a>
<a class="sourceLine" id="cb3-11" data-line-number="11"><span class="kw">rm</span>(taxa_abund)</a></code></pre></div>
</div>
<div id="gene-family-data" class="section level4">
<h4><span class="header-section-number">2.1.1.3</span> Gene family data</h4>
<p>Gene family and pathway abundance data have already been normalized to an RPKM-like metric in these tables (read counts or pathway copies per kilobase length per million reads). This normalization approach does not address compositional bias in the data, but for the sake of simplicity we will proceed with these quantities.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Read in file, get sample subsets, rename columns, split row IDs into EC number and Taxon, filter features present in less than 3 samples, save</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">ecs &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/ecs_3.tsv.gz&quot;</span>)</a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="cf">for</span>(site <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;MGH&quot;</span>, <span class="st">&quot;Cincinnati&quot;</span>)){</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">  site_samples &lt;-<span class="st"> </span><span class="kw">filter</span>(subject_metadata_mgh_cinc, site_name <span class="op">==</span><span class="st"> </span>site)<span class="op">$</span>SampleID</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">  ecs1 &lt;-<span class="st"> </span>ecs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="kw">any_of</span>(<span class="kw">c</span>(<span class="kw">names</span>(ecs)[<span class="dv">1</span>], <span class="kw">paste0</span>(site_samples, <span class="st">&quot;_level4ec&quot;</span>))))</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">  <span class="kw">colnames</span>(ecs1) =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Feature&quot;</span>, <span class="kw">gsub</span>(<span class="st">&quot;_level4ec&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">names</span>(ecs1)[<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(ecs1)]))</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">  ecs1 &lt;-<span class="st"> </span><span class="kw">mutate</span>(ecs1, <span class="dt">EC =</span> <span class="kw">gsub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|.*&quot;</span>, <span class="st">&quot;&quot;</span>, Feature), <span class="dt">Taxon =</span> <span class="kw">gsub</span>(<span class="st">&quot;.*</span><span class="ch">\\</span><span class="st">|&quot;</span>, <span class="st">&quot;&quot;</span>, Feature))</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">  ec_samps &lt;-<span class="st"> </span><span class="kw">apply</span>(<span class="kw">select</span>(ecs1, <span class="op">!</span>Feature <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span>Taxon <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span>EC), <span class="dv">1</span>, <span class="cf">function</span>(x){ <span class="kw">length</span>(x[x <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>])})</a>
<a class="sourceLine" id="cb4-10" data-line-number="10">  ecs_keep &lt;-<span class="st"> </span><span class="kw">which</span>(ec_samps <span class="op">&gt;=</span><span class="st"> </span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb4-11" data-line-number="11">  ecs1 &lt;-<span class="st"> </span>ecs1[ecs_keep,]</a>
<a class="sourceLine" id="cb4-12" data-line-number="12">  <span class="kw">write_tsv</span>(ecs1, <span class="dt">path =</span> <span class="kw">paste0</span>(<span class="st">&quot;data/ECnumber_profile_&quot;</span>, site, <span class="st">&quot;.tsv&quot;</span>))</a>
<a class="sourceLine" id="cb4-13" data-line-number="13">}</a>
<a class="sourceLine" id="cb4-14" data-line-number="14"></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="kw">rm</span>(ecs)</a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="kw">rm</span>(ecs1)</a></code></pre></div>
</div>
<div id="pathway-data" class="section level4">
<h4><span class="header-section-number">2.1.1.4</span> Pathway data</h4>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># Read in file, get sample subsets, rename columns, split row IDs into Pathway and Taxon, filter features present in less than 3 samples, save</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">pathway_data &lt;-<span class="st"> </span><span class="kw">read_tsv</span>(<span class="st">&quot;data/pathabundances_3.tsv.gz&quot;</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="cf">for</span>(site <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;MGH&quot;</span>, <span class="st">&quot;Cincinnati&quot;</span>)){</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">  site_samples &lt;-<span class="st"> </span><span class="kw">filter</span>(subject_metadata_mgh_cinc, site_name <span class="op">==</span><span class="st"> </span>site)<span class="op">$</span>SampleID</a>
<a class="sourceLine" id="cb5-7" data-line-number="7">  pathways1 &lt;-<span class="st"> </span>pathway_data <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="kw">any_of</span>(<span class="kw">c</span>(<span class="kw">names</span>(pathway_data)[<span class="dv">1</span>], <span class="kw">paste0</span>(site_samples, <span class="st">&quot;_pathabundance_cpm&quot;</span>))))</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">  <span class="kw">colnames</span>(pathways1) =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Feature&quot;</span>, <span class="kw">gsub</span>(<span class="st">&quot;_pathabundance_cpm&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">names</span>(pathways1)[<span class="dv">2</span><span class="op">:</span><span class="kw">ncol</span>(pathways1)]))</a>
<a class="sourceLine" id="cb5-9" data-line-number="9">  pathways1 &lt;-<span class="st"> </span><span class="kw">mutate</span>(pathways1, <span class="dt">Pathway =</span> <span class="kw">gsub</span>(<span class="st">&quot;</span><span class="ch">\\</span><span class="st">|.*&quot;</span>, <span class="st">&quot;&quot;</span>, Feature), <span class="dt">Taxon =</span> <span class="kw">gsub</span>(<span class="st">&quot;.*</span><span class="ch">\\</span><span class="st">|&quot;</span>, <span class="st">&quot;&quot;</span>, Feature))</a>
<a class="sourceLine" id="cb5-10" data-line-number="10">  pathways1 &lt;-<span class="st"> </span><span class="kw">pivot_longer</span>(pathways1, <span class="dt">cols =</span> <span class="kw">names</span>(pathways1)[<span class="dv">2</span><span class="op">:</span>(<span class="kw">ncol</span>(pathways1)<span class="op">-</span><span class="dv">2</span>)], <span class="dt">names_to =</span> <span class="st">&quot;Sample&quot;</span>)</a>
<a class="sourceLine" id="cb5-11" data-line-number="11">  bad_pathways &lt;-<span class="st"> </span>pathways1 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(value <span class="op">!=</span><span class="st"> </span><span class="dv">0</span> <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(value)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count</span>(Feature) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(n <span class="op">&lt;</span><span class="st"> </span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb5-12" data-line-number="12">  pathways1 &lt;-<span class="st"> </span>pathways1 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(<span class="op">!</span>Feature <span class="op">%in%</span><span class="st"> </span>bad_pathways<span class="op">$</span>Feature) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pivot_wider</span>(<span class="dt">id_cols =</span> <span class="kw">c</span>(<span class="st">&quot;Feature&quot;</span>, <span class="st">&quot;Pathway&quot;</span>, <span class="st">&quot;Taxon&quot;</span>), <span class="dt">names_from =</span> <span class="st">&quot;Sample&quot;</span>, <span class="dt">values_from =</span> <span class="st">&quot;value&quot;</span>)</a>
<a class="sourceLine" id="cb5-13" data-line-number="13">  <span class="kw">write_tsv</span>(pathways1, <span class="dt">path =</span> <span class="kw">paste0</span>(<span class="st">&quot;data/pathway_profile_&quot;</span>, site, <span class="st">&quot;.tsv&quot;</span>))</a>
<a class="sourceLine" id="cb5-14" data-line-number="14">}</a>
<a class="sourceLine" id="cb5-15" data-line-number="15"></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="kw">rm</span>(pathway_data)</a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="kw">rm</span>(pathways1)</a></code></pre></div>
</div>
<div id="metabolite-data" class="section level4">
<h4><span class="header-section-number">2.1.1.5</span> Metabolite data</h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># Read in file, get sample subsets, filter features present in less than 3 samples, save</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"></a>
<a class="sourceLine" id="cb6-3" data-line-number="3">metabolites &lt;-<span class="st"> </span><span class="kw">read_csv</span>(<span class="st">&quot;data/HMP2_metabolomics.csv.gz&quot;</span>)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4"><span class="cf">for</span>(site <span class="cf">in</span> <span class="kw">c</span>(<span class="st">&quot;MGH&quot;</span>, <span class="st">&quot;Cincinnati&quot;</span>)){</a>
<a class="sourceLine" id="cb6-5" data-line-number="5">  site_samples &lt;-<span class="st"> </span><span class="kw">filter</span>(subject_metadata_mgh_cinc, site_name <span class="op">==</span><span class="st"> </span>site <span class="op">&amp;</span><span class="st"> </span>data_type <span class="op">==</span><span class="st"> &quot;metabolomics&quot;</span>)<span class="op">$</span>SampleID</a>
<a class="sourceLine" id="cb6-6" data-line-number="6">  metabolites1 &lt;-<span class="st"> </span>metabolites <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="kw">any_of</span>(<span class="kw">c</span>(<span class="kw">names</span>(metabolites)[<span class="dv">1</span><span class="op">:</span><span class="dv">7</span>], site_samples))) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="st">    </span><span class="kw">pivot_longer</span>(metabolites1, <span class="dt">cols =</span> <span class="kw">names</span>(metabolites1)[<span class="dv">8</span><span class="op">:</span><span class="kw">ncol</span>(metabolites1)], <span class="dt">names_to =</span> <span class="st">&quot;Sample&quot;</span>)</a>
<a class="sourceLine" id="cb6-8" data-line-number="8">  <span class="co"># We will filter metabolite features present in very few samples</span></a>
<a class="sourceLine" id="cb6-9" data-line-number="9">  bad_metabolites &lt;-<span class="st"> </span>metabolites1 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(value <span class="op">!=</span><span class="st"> </span><span class="dv">0</span> <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(value)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">count</span>(Compound) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(n <span class="op">&lt;</span><span class="st"> </span><span class="dv">3</span>)</a>
<a class="sourceLine" id="cb6-10" data-line-number="10">  metabolites1 &lt;-<span class="st"> </span>metabolites1 <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(<span class="op">!</span>Compound <span class="op">%in%</span><span class="st"> </span>bad_metabolites<span class="op">$</span>Compound) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pivot_wider</span>(<span class="dt">names_from =</span> <span class="st">&quot;Sample&quot;</span>, <span class="dt">values_from =</span> <span class="st">&quot;value&quot;</span>)</a>
<a class="sourceLine" id="cb6-11" data-line-number="11">  <span class="kw">write_tsv</span>(metabolites1, <span class="dt">path =</span> <span class="kw">paste0</span>(<span class="st">&quot;data/metabolite_profile_&quot;</span>, site, <span class="st">&quot;.tsv&quot;</span>))</a>
<a class="sourceLine" id="cb6-12" data-line-number="12">}</a>
<a class="sourceLine" id="cb6-13" data-line-number="13"></a>
<a class="sourceLine" id="cb6-14" data-line-number="14"><span class="kw">rm</span>(metabolites)</a>
<a class="sourceLine" id="cb6-15" data-line-number="15"><span class="kw">rm</span>(metabolites1)</a>
<a class="sourceLine" id="cb6-16" data-line-number="16"><span class="co"># ## Various MS methods used</span></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"><span class="co"># metabolites %&gt;% group_by(Method) %&gt;% summarize(nFeat = length(unique(Compound)))</span></a></code></pre></div>
</div>
</div>
</div>
<div id="option-2-download-smaller-pre-processed-datasets" class="section level2">
<h2><span class="header-section-number">2.2</span> Option 2: Download smaller pre-processed datasets</h2>
<p>If needed, you can download smaller pre-filtered versions of the tables at the links below. Save these archives in your R project folder and uncompress them.</p>
<ul>
<li><a href="downloads/IBD_metagenomics.tar.gz">Metagenomics data</a></li>
<li><a href="downloads/IBD_metabolomics.tar.gz">Metabolomics data</a></li>
<li><a href="downloads/IBD_metadata.tar.gz">Metadata</a></li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">download.file</span>(<span class="st">&quot;https://cnoecker.github.io/MultiOmicsTutorial/downloads/IBD_metagenomics.tar.gz&quot;</span>, <span class="dt">destfile =</span> <span class="st">&quot;IBD_metagenomics.tar.gz&quot;</span>)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2"><span class="kw">untar</span>(<span class="st">&quot;IBD_metagenomics.tar.gz&quot;</span>)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3"></a>
<a class="sourceLine" id="cb7-4" data-line-number="4"><span class="kw">download.file</span>(<span class="st">&quot;https://cnoecker.github.io/MultiOmicsTutorial/downloads/IBD_metabolomics.tar.gz&quot;</span>, <span class="dt">destfile =</span> <span class="st">&quot;IBD_metabolomics.tar.gz&quot;</span>)</a>
<a class="sourceLine" id="cb7-5" data-line-number="5"><span class="kw">untar</span>(<span class="st">&quot;IBD_metabolomics.tar.gz&quot;</span>)</a>
<a class="sourceLine" id="cb7-6" data-line-number="6"></a>
<a class="sourceLine" id="cb7-7" data-line-number="7"><span class="kw">download.file</span>(<span class="st">&quot;https://cnoecker.github.io/MultiOmicsTutorial/downloads/IBD_metadata.tar.gz&quot;</span>, <span class="dt">destfile =</span> <span class="st">&quot;IBD_metadata.tar.gz&quot;</span>)</a>
<a class="sourceLine" id="cb7-8" data-line-number="8"><span class="kw">untar</span>(<span class="st">&quot;IBD_metadata.tar.gz&quot;</span>)</a></code></pre></div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
